### ISSUE TITLE
question: Exporting plot data

### ISSUE THREAD

**Issue by wyessen** [facebookresearch/visdom#670]
I know there are a few existing issues. I am wondering, as of right now, what is the most hygienic way of exporting plot data from an instance of the Visdom server? Ideally in some sort of image format, but if not, some kind of format that I could use to recreate the images on the client side?

**Comment by JackUrb** [comment]
Hi @wyessen - This has been on the radar since at least #232, but there really isn't any great way to automatically get the plotly plots out as images as it would be breaking the Terms of Service to allow such a behavior. Without direct access to the filesystem of the machine hosting the visdom server, the closest thing would be being able to export the environment `.json` and being able to run a local server later using that pre-saved `json`, but this hasn't yet been enabled. It's planned for shortly after the `0.2` release.

**Comment by wyessen** [comment]
@JackUrb Thanks. Actually, here is a scenario: let's say I'm running Visdom on a server with direct access to the server's filesystem. Let's say I have a python script that plots to Visdom. This script also knows the `env_path` that Visdom is using (i.e. it knows where Visdom keeps its working files). Now after the script is done plotting, the script wants to grab the Visdom data (in whatever format), and save it in some other location on the FS. 

How can we achieve this? Keep in mind, we don't have direct access to the web UI. This relates to https://github.com/facebookresearch/visdom/issues/591

**Comment by wyessen** [comment]
@JackUrb For instance, there is the `vis.get_window_data` method which I could potentially use to get the window data. But what format is this data in? How can I reproduce the image from this data?

**Comment by JackUrb** [comment]
If you can access the server but not the web ui, you'll first want to call `vis.save('env_name')` on the environment you're plotting to (which is default `'main'`) once the experiment is done. You can then navigate to the default env location on the server:
https://github.com/facebookresearch/visdom/blob/beb9fa94e5f885c34d7119571133d1d5157e84ea/py/visdom/server.py#L47
Copy over the `main.json` file to your local device into the local device's default visdom env location. Then you can launch the visdom server on your local device and view the contents of the environment from localhost.

**Comment by wyessen** [comment]
@JackUrb Fair enough, thanks. I also see that Visdom has (as you mentioned above) environments and views, etc, which can be used to organize data (perhaps by project). For instance, I imagine a script that is plotting on Visdom could create its own environment.

So the question is, what is the API for doing this? e.g. something like `vis.create_environment(env_name)`?

**Comment by JackUrb** [comment]
Envs are automatically created when you plot to one that doesn't exist yet, so the first call to 
```
vis.line(..., env='my_super_cool_experiment')
```
will create a local state in the server for the environment `"my_super_cool_experiment"`, and then future calls such as
```
vis.scatter(..., env='my_super_cool_experiment')
```
will put the new plots into the same organizational place. Afterwards if you wanted to get the state to be saved to disk, you would call
```
vis.save(['my_super_cool_experiment'])
```
and then the local state for that environment is saved to disk at `DEFAULT/ENV/LOCATION/my_super_cool_experiment.json`.

**Comment by wyessen** [comment]
@JackUrb Excellent, thank you! Is it possible to configure Visdom to purge stale environments after a specified time? Or do I have to do it manually (i.e. install a watchdog script that will periodically purge stale environments with `vis.delete_env()`?

**Comment by JackUrb** [comment]
You'll have to manage your own purges. At the moment we don't have anything to check for environment staleness, but it would be a cool contribution if you happen to automate it in some generalizable way!

**Comment by ravikiranrao** [comment]
> Envs are automatically created when you plot to one that doesn't exist yet, so the first call to
> 
> ```
> vis.line(..., env='my_super_cool_experiment')
> ```
> 
> will create a local state in the server for the environment `"my_super_cool_experiment"`, and then future calls such as
> 
> ```
> vis.scatter(..., env='my_super_cool_experiment')
> ```
> 
> will put the new plots into the same organizational place. Afterwards if you wanted to get the state to be saved to disk, you would call
> 
> ```
> vis.save(['my_super_cool_experiment'])
> ```
> 
> and then the local state for that environment is saved to disk at `DEFAULT/ENV/LOCATION/my_super_cool_experiment.json`.


Thanks @JackUrb. Sorry, although this is an old thread I found this to the the relevant one for my question.

Is there a way to change the default save path from `DEFAULT/ENV/LOCATION/my_super_cool_experiment.json` to something of our choice? One workaround is to call subprocess to copy the json. Any default way?

Thanks.

**Comment by da-h** [comment]
Hey @ravikiranrao!

Do you mean to change the path *besides* the option to change the working directory for all environments using the `-env_path OTHER/DEFAULT/PATH` argument?
**If yes, could you please give a scenario for this feature?**  
(What is the used case for such a feature?)

---

**Direct Solution Idea**:  
One solution for your query, albeit a bit hacky, could be to create a filesystem symlink in the working directory to the path you would like the data to be saved to.
For instance, let's say you are using the default env-path `~/,visdom` but would like to have one specific experiment named `theoneexp` to be saved somewhere else, say `/tmp/importantexp.json`.

In that case, just create the symlink beforehand:
```sh
ln -s /tmp/importantexp.json ~/.visdom/theoneexp.json
```

After that, writing to the env `theoneexp` would be saved in `/tmp/importantexp.json`.
For instance using this code-snippet:
```python
env = "theoneexp"
viz.text("Hello World!", env=env)
viz.save([env])
```

Best,
*da-h*

**Comment by ravikiranrao** [comment]
Thanks @da-h for the reply!

> Do you mean to change the path besides the option to change the working directory for all environments using the -`env_path OTHER/DEFAULT/PATH` argument?
**If yes, could you please give a scenario for this feature?**
(What is the used case for such a feature?)

Not really the working directory for all environments, rather just `.json' files of my interest.

Suppose we have an environment `my_env`. Currently, if I want to save (in whatever format) *all* the plots from `my_env`, either I can click `folder icon` (manage environment), then click `save`. This saves the `.json` to `~/.visdom` by default, OR using this code snippet inside my python program
```
env = "my_env"
viz.line(Y=..., X=..., env=env)
viz.save([my_env])
```

Rather what I am looking for is 
```
viz.save([my_env], path_to_dir)
```
which saves a *copy* of `my_env.json` to `path_to_dir`.

Currently, I am doing this in two steps: 1) save `my_env` to `~/.visdom` 2) copy `my_env.json` to `path_to_dir` using `shutil`. Something like this:

```
env = "my_env"
viz.line(Y=..., X=..., env=env)
viz.save([my_env])

shutil.copy2(os.path.expanduser("~/.visdom/my_env.json"), path_to_dir)
```

In summary, I was looking for a feature something like this: `viz.save([my_env], path_to_dir)`.

Best,
Ravi

**Comment by JackUrb** [comment]
Sorry for the delay here @ravikiranrao - I'd be up for this being a feature in visdom. Feel free to open a PR that adds this duplicate save functionality to `Visdom.save`!
